llvm: provide fix for python dependencies

robertu94 opened the pull request:
Previously compiler-rt didn't correctly passthrough cmake
variables for python when building the various santizers.
This patch passes these variables through.

Fixes #19908
See also: https://bugs.llvm.org/show_bug.cgi?id=48180

robertu94 committed:
llvm: provide fix for python dependencies

Previously compiler-rt didn't correctly passthrough cmake
variables for python when building the various santizers.
This patch passes these variables through.

This patch may also correctly apply to any version of LLVM
to any version of LLVM that uses the newer monorepo style organization,
and any older llvm newer than 7.0.0 as long as the paths were set
appropriately.  However, this was not done so because it was not
tested with older LLVM releases.

Fixes #19908
See also: https://bugs.llvm.org/show_bug.cgi?id=48180

robertu94 commented:
@naromero77 @trws  Sorry for the delay, things have been busy recently.  I looked at the llvm git history, and the python patch will probably apply to any version of LLVM that uses the newer monorepo style organization, and any older llvm newer than 7.0.0 as long as the paths were set appropriately.  If someone else wants to test with earlier versions, we could probably push the patch lower bound back.

I think the flake8 failures are before my change.

naromero77 commented:
You can do `spack flake8` to see the failures.

Here are the lines that are failing the flake8 test (from travis-CI):
```
Modified files:
  var/spack/repos/builtin/packages/llvm/package.py
=======================================================
var/spack/repos/builtin/packages/llvm/package.py:384: [E501] line too long (84 > 79 characters)
var/spack/repos/builtin/packages/llvm/package.py:386: [E501] line too long (84 > 79 characters)
```

LLVM will do a 11.0.1 release next, so the patch shouldn't only be applied to LLVM releases after 11.0.0:

```
    patch("llvm_python_path.patch", when="@11.0.0")
```

If you think the patch applies to versions of LLVM since the monorepo, go ahead and include those in the patch line. I would not bother patching versions of LLVM prior to the monorepo.

robertu94 commented:
@naromero77 I've done some investigation, and I think the patch was added after a LLVM 11.0.1 feature freeze because none of the current 11.0.1-rc branches contain the commit with the fix.  I could be wrong since I'm not that familar with LLVM and this was my first patch to LLVM.  I've changed the spack  patch line per your request, and we can update this later if need be.

```console
[runderwood@voyager llvm-project]$ git name-rev --tags 16de50895e96adbe261a5ce2498366bda7b3fccd
16de50895e96adbe261a5ce2498366bda7b3fccd undefined
```

I've fixed the flake8 problems.

naromero77 commented:
@trws Are you OK with incorporating this patch?

trws commented:
I'm fine with this so long as the only version that needs it is the `11.0.0` version.

naromero77 commented:
LGTM

naromero77 commented:
@alalazo

adamjstewart commented on var/spack/repos/builtin/packages/llvm/package.py:
```diff
@@ -365,16 +369,24 @@ def setup_run_environment(self, env):
         ]
 
+        if python.version >= Version("3.0.0"):
+            cmake_args.append("-DPython3_EXECUTABLE={}".format(
+                              python.command.path))
+        else:
+            cmake_args.append("-DPython2_EXECUTABLE={}".format(
```
```suggestion
            cmake_args.append("-DPython3_EXECUTABLE={0}".format(
                              python.command.path))
        else:
            cmake_args.append("-DPython2_EXECUTABLE={0}".format(
```
Needed for Python 2.6 support

robertu94 replied:
@adamjstewart done
